runtime/lua/vim/lsp/completion.lua

```diff
@@ -0,0 +1,128 @@
+local M = {}
+
+--- @class vim.lsp.completion.BufHandle
+--- @field clients table<integer, lsp.Client>
+--- @field triggers table<string, lsp.Client[]>
+
+--- @type table<integer, vim.lsp.completion.BufHandle>
+local buf_handles = {}
+
+--- @param handle vim.lsp.completion.BufHandle
+local function insert_char_pre_cb(handle) end
+
+local function text_changed_p_cb() end
+
+local function text_changed_i_cb() end
+
+local function insert_leave_cb() end
+
+local function complete_done_cb() end
+
+--- Initializes the completion commands for the given client.
+---
+--- @param client lsp.Client
+local function init_commands(client)
+  local trigger_completion_cmd = 'editor.action.triggerSuggest'
+
+  -- Check if the command is in the global registry or in the client's commands.
+  if
+    not vim.lsp.commands[trigger_completion_cmd] and not client.commands[trigger_completion_cmd]
+  then
+    client.commands[trigger_completion_cmd] = function()
+      pcall(M.trigger_completion)
+    end
+  end
+end
+
+function M.trigger_completion() end
+
+--- TODO(mariasolos): Document.
+---
+--- @param client_id integer Client ID
+--- @param bufnr integer Buffer handle, or 0 for the current buffer
+function M.attach(client_id, bufnr)
+  bufnr = (bufnr == 0 and vim.api.nvim_get_current_buf()) or bufnr
+
+  if not buf_handles[bufnr] then
+    buf_handles[bufnr] = { clients = {}, triggers = {} }
+
+    -- Attach to buffer events.
+    vim.api.nvim_buf_attach(bufnr, false, {
+      on_detach = function(_, buf)
+        buf_handles[buf] = nil
+      end,
+      on_reload = function(_, buf)
+        M.attach(client_id, buf)
+      end,
+    })
+
+    -- Set up autocommands.
+    local group =
+      vim.api.nvim_create_augroup(string.format('vim/lsp/completion-%d', bufnr), { clear = true })
+    vim.api.nvim_create_autocmd('InsertCharPre', {
+      group = group,
+      buffer = bufnr,
+      callback = function()
+        insert_char_pre_cb(buf_handles[bufnr])
+      end,
+    })
+    vim.api.nvim_create_autocmd('TextChangedP', {
+      group = group,
+      buffer = bufnr,
+      callback = function()
+        text_changed_p_cb()
+      end,
+    })
+    vim.api.nvim_create_autocmd('TextChangedI', {
+      group = group,
+      buffer = bufnr,
+      callback = function()
+        text_changed_i_cb()
+      end,
+    })
```

Comment:
These are only needed when we want to trigger completion when deleting. `lsp_compl` makes this configurable, but I think we should just always enable it here (?). WDYT?